File: app/src/main/java/org/oppia/app/topic_index/TopicIndexFragment.java
package org.oppia.app.topic_index;

import android.os.Bundle;
import android.util.Log;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import org.oppia.app.backend.model.TopicIndexModel;
import org.oppia.app.backend.model.TopicIndexHandler;
import org.oppia.app.R;
import org.oppia.app.backend.OppiaGaeClient;
import org.oppia.app.backend.api.TopicIndexHandlerService;
import org.oppia.app.topic_index.adapter.TopicIndexAdapter;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import java.util.ArrayList;
import java.util.List;
/** This Fragment contains a list of Topics summary */
public class TopicIndexFragment extends Fragment {

  RecyclerView rvTopicSummary;

  List<TopicIndexModel> topicSummaryList = new ArrayList<>();
  GridLayoutManager gridLayoutManager;

  TopicIndexAdapter topicIndexAdapter;
  private static final String TAG = TopicIndexFragment.class.getName();
  public TopicIndexFragment() {
    // Required empty public constructor
  }


  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
    View view =  inflater.inflate(R.layout.topic_index_fragment, container, false);
    rvTopicSummary = view.findViewById(R.id.rvTopicSummary);
    fetchTopicSummary();
    initViews();

    return view;

  }

  private void initViews() {
    gridLayoutManager = new GridLayoutManager(getActivity(), 2);
    rvTopicSummary.setLayoutManager(gridLayoutManager);
    topicIndexAdapter = new TopicIndexAdapter(topicSummaryList, getActivity());
    rvTopicSummary.setAdapter(topicIndexAdapter);
  }

  private void fetchTopicSummary() {

    TopicIndexHandlerService appStoreApiService = OppiaGaeClient.INSTANCE.getRetrofitInstance().create(TopicIndexHandlerService.class);

    Call<TopicIndexHandler> getStoresResponseCall = appStoreApiService.getTopicIndex();

    getStoresResponseCall.enqueue(new Callback<TopicIndexHandler>() {
      @Override
      public void onResponse(Call<TopicIndexHandler> call, Response<TopicIndexHandler> response) {


        Integer code = response.code();
        switch (code) {
          case 200:
          case 201:
            TopicIndexHandler topicSummaryHandler =  response.body();

            topicSummaryList.addAll(topicSummaryHandler.getTopic_summary_dicts());
            Log.d(TAG, "Topic Name: ******************" +response.body());

            topicIndexAdapter.notifyDataSetChanged();
            break;
        }

      }

      @Override
      public void onFailure(Call<TopicIndexHandler> call, Throwable t) {
        Log.d(TAG,"Failure ****************"+ t.getMessage());



      }
    });


  }

}

File: app/src/main/java/org/oppia/app/topic_index/TopicIndexFragment.kt
package org.oppia.app.topic_index

import android.os.Bundle
import android.util.Log
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.oppia.app.backend.model.TopicIndexModel
import org.oppia.app.backend.model.TopicIndexHandler
import org.oppia.app.R
import org.oppia.app.topic_index.adapter.TopicIndexAdapter
import org.oppia.data.backends.gae.OppiaGaeClient
import org.oppia.data.backends.gae.api.TopicIndexHandlerService
import retrofit2.Call

import retrofit2.Callback
import retrofit2.Response

import java.util.ArrayList

/** This Fragment contains a list of Topics summary  */
class TopicIndexFragment : Fragment() {

   var rvTopicindex: RecyclerView? = null

  internal var topicSummaryList: MutableList<TopicIndexModel> = ArrayList()
   var gridLayoutManager: GridLayoutManager? = null

  var topicIndexAdapter: TopicIndexAdapter? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

  }

  override fun onCreateView(
    inflater: LayoutInflater, container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    val view = inflater.inflate(R.layout.topic_index_fragment, container, false)

     rvTopicindex = view.findViewById(R.id.rvTopicSummary) as RecyclerView

    initViews()

    fetchTopicSummary()

    return view

  }

  private fun initViews() {
    gridLayoutManager = GridLayoutManager(activity, 2)
    rvTopicindex!!.layoutManager = gridLayoutManager

    topicIndexAdapter = TopicIndexAdapter(topicSummaryList, this)
    rvTopicindex!!.adapter = topicIndexAdapter
  }

  private fun fetchTopicSummary() {

    val topicIndexService: TopicIndexHandlerService =
      OppiaGaeClient.retrofitInstance!!.create(TopicIndexHandlerService::class.java)

    val call: Call<TopicIndexHandler> = topicIndexService.getTopicIndex()

    call.enqueue(object : Callback<TopicIndexHandler> {

      override fun onResponse(call: Call<TopicIndexHandler>?, response: Response<TopicIndexHandler>?) {

//        val code = response!!.code()
//        when (code) {
//          200, 201 -> {
            val topicSummaryHandler = response!!.body()

            topicSummaryList.addAll(topicSummaryHandler!!.topic_summary_dicts!!)
            Log.d(TAG, "Topic Name: ******************" + response.body())

            topicIndexAdapter!!.notifyDataSetChanged()
//          }
//        }

      }

      override fun onFailure(call: Call<TopicIndexHandler>?, t: Throwable?) {

        Log.e(TAG, t.toString())

      }

    })

//    val appStoreApiService = OppiaGaeClient.retrofitInstance!!.create(TopicIndexHandlerService::class.java)
//
//    val getStoresResponseCall = appStoreApiService.getTopicIndex()
//
//    getStoresResponseCall.enqueue(object : SortedList.Callback<TopicIndexHandler>() {
//      fun onResponse(call: Call<TopicIndexHandler>, response: Response<TopicIndexHandler>) {
//
//        val code = response.code()
//        when (code) {
//          200, 201 -> {
//            val topicSummaryHandler = response.body()
//
//            topicSummaryList.addAll(topicSummaryHandler.topic_summary_dicts!!)
//            Log.d(TAG, "Topic Name: ******************" + response.body())
//
//            topicIndexAdapter.notifyDataSetChanged()
//          }
//        }
//
//      }
//
//      fun onFailure(call: Call<TopicIndexHandler>, t: Throwable) {
//        Log.d(TAG, "Failure ****************" + t.message)
//
//      }
//    })

  }

  companion object {
    private val TAG = TopicIndexFragment::class.java.name
  }

}// Required empty public constructor
